src/scheduler/workers/helpers.annualReview.ts

```diff
@@ -0,0 +1,87 @@
+import { AnnualReviewKeyDates, Audit, List } from "server/models/types";
+import { ListItemWithHistory } from "server/components/dashboard/listsItems/types";
+import { createAnnualReviewProviderUrl, formatDate, processPostEmail, processPostProviderEmail } from "./helpers";
+import { sendAnnualReviewProviderEmail } from "server/services/govuk-notify";
+import { BaseDeserialisedWebhookData } from "server/models/listItem/providers/deserialisers/types";
+import { lowerCase, startCase } from "lodash";
+import { endOfDay, subDays } from "date-fns";
+
+export async function emailProviderForAnnualReviewKeyDates(list: List, listItem: ListItemWithHistory) {
+  const annualReviewProviderUrl = createAnnualReviewProviderUrl(listItem);
+  const unpublishDate = new Date(list.jsonData.currentAnnualReview?.keyDates.unpublished.UNPUBLISH ?? "");
+  return await sendAnnualReviewProviderEmail(
+    (listItem.jsonData as BaseDeserialisedWebhookData).emailAddress,
+    lowerCase(startCase(listItem.type)),
+    list?.country?.name ?? "",
+    (listItem.jsonData as BaseDeserialisedWebhookData).contactName,
+    formatDate(unpublishDate),
+    annualReviewProviderUrl
+  );
+}
+
+export async function processAnnualReviewEmails(
+  list: List,
+  listItemsForList: ListItemWithHistory[],
+  audit: Audit,
+  today: Date,
+  annualReviewKeyDates: AnnualReviewKeyDates
+): Promise<boolean> {
+  if (
+    await processPostEmail(
+      list,
+      audit,
+      "POST_ONE_MONTH",
+      today,
+      new Date(annualReviewKeyDates?.POST_ONE_MONTH ?? ""),
+      subDays(endOfDay(new Date(annualReviewKeyDates?.POST_ONE_WEEK ?? "")), 1),
+      "sendOneMonthPostEmail"
+    )
+  ) {
+    return true;
+  }
+
+  if (
+    await processPostEmail(
+      list,
+      audit,
+      "POST_ONE_WEEK",
+      today,
+      new Date(annualReviewKeyDates?.POST_ONE_WEEK ?? ""),
+      subDays(endOfDay(new Date(annualReviewKeyDates?.POST_ONE_DAY ?? "")), 1),
+      "sendOneWeekPostEmail"
+    )
+  ) {
+    return true;
+  }
+
+  if (
+    await processPostEmail(
+      list,
+      audit,
+      "POST_ONE_DAY",
+      today,
+      new Date(annualReviewKeyDates?.POST_ONE_DAY ?? ""),
+      endOfDay(new Date(annualReviewKeyDates?.POST_ONE_DAY ?? "")),
+      "sendOneDayPostEmail"
+    )
+  ) {
+    return true;
+  }
+
+  if (
+    await processPostProviderEmail(
+      list,
+      listItemsForList,
+      audit,
+      "START",
+      today,
+      new Date(annualReviewKeyDates?.START ?? ""),
+      endOfDay(new Date(annualReviewKeyDates?.START ?? "")),
+      "sendStartedPostEmail",
+      "sendStartedProviderEmail"
+    )
+  ) {
+    return true;
+  }
+  return false;
+}
```

Comment:
I don't think I agree that a function named "processPostEmail" should control logic/flow.

"if the dates are within these dates, process the emails for post".

I'm also not sure why boolean is being returned.. it looks like processPostEmail should already be returning a boolean (since you are using it in an if statement).. wouldn't it make more sense to just return the promise? 

what happened to my suggestion for destructing the dates also?

does today (Date) need to be passed in?

Also, if there are more than 3 parameters I would consider refactoring. the function is doing "too much". It makes testing unwieldy when you are trying to manage 9 parameters. It is also unclear what the function is doing if I don't step into it. 

I know we want to get this out ASAP. But we also want it to be debuggable/fixable by everyone in 3 months time.


please use this suggestion to tidy the code. There shouldn't be much for you to do apart from copy+paste this and verify it works. This was from the previous PR, but I realise you may need to use the opposite end/start of day. e.g. use endOfDay(startOfRange), startOfDay(endOfRange). 



```suggestion
import { isWithinInterval, startOfDay, endOfDay, Interval, isToday } from "date-fns"

// you may need to change the interval type to accept string. 

export function isTodayWithinInterval(interval: Interval) {
  const { start, end } = interval;
  const today = getTodayDate();

  if(!interval.start || !interval.end) {
    // log or throw, then return
    throw new Error("interval start or end date was empty")
   }
   
   let startDate, endDate;
   
   try {
     startDate = new Date(start);
     endDate = new Date(endDate);
   } catch (e) {
      throw new Error(`isTodayWithinInterval could not parse ${interval} to date, ${e}`)
   
   }
  
  const intervalWithStartAndEndTime = {
     start: endOfDay(startDate), // date:23:59:599
     end: startOfDay(endDate), // date:00:00
  }
   
  return isWithinInterval(today, intervalWithStartAndEndTime);
}


export async function processAnnualReviewEmails(
  list: List,
  listItemsForList: ListItemWithHistory[],
  audit: Audit,
  today: Date,
  annualReviewKeyDates: AnnualReviewKeyDates
): Promise<boolean> {
  const {
    POST_ONE_MONTH = "", // removes the need to inline ?? "" for each function call
    POST_ONE_WEEK = "",
    POST_ONE_DAY = "",
    START = ""
  } = annualReviewKeyDates;
  
  if(isBefore(today, POST_ONE_MONTH)) {
    // if you want to handle if today is before post_one_month
  }

  if (
    isTodayWithinInterval({ start: POST_ONE_MONTH, end: POST_ONE_WEEK })
  ) {
     // send the 1 month email if today is
     // - at least 1 week before the start date at 23:59:599
     // - at most 1 month before the start date at 00:00
     // i.e. this will attempt to run every day between these two dates.
     // Also, I think the most needed to process (and send) emails, is the list and the type of reminder you want to send. 
     return await processPostProviderEmail(list, "POST_ONE_MONTH");
  }
 

  if (
    isTodayWithinInterval({ start: POST_ONE_WEEK, end: POST_ONE_DAY })
  ) {
    // between POST_ONE_WEEK at 23:59:599 and POST_ONE_DAY 00:00:00 
    return await processPostProviderEmail(list, "POST_ONE_WEEK");
  }
 
  
  if(isToday(POST_ONE_DAY)) {
    return await processPostProviderEmail(list, "POST_ONE_DAY");
  }
  
  if(isToday(START)) {
  // you may also want to handle this as a range. if(isPast(START))
    logger.info(`${START} is today`);
    return await processPostProviderEmail(list, "POST_START");
  }
  
  logger.info("no valid action found for list id ${list.id}")
  
}
```